design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,300 @@
+
+- A generic S3 user API is not defined in this proposal but should be given serious consideration.  A user API should allow cluster users to generate new S3 users, tenants, and access keys in an object store.  This API could be used as method of manually creating object store users or incorporated into the bucket provisioning automation to generate the user's access key.
+
+---
+
+## API Specifications
+
+### OBC Custom Resource Definition
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: objectbucketclaims.rook.io
+spec:
+  group: rook.io
+  version: v1
+  scope: namespaced
+  names:
+      kind: ObjectBucketClaim
+      listKind: ObjectBucketClaimList
+      singular: objectbucketclaim
+      plural: objectbucketclaims
+```
+
+
+### OBC Custom Resource (User Defined)
+
+```yaml
+apiVersion: rook.io/v1
+kind: ObjectBucketClaim
+metadata:
+  name: MY-BUCKET-1 [1]
+  namespace: USER-NAMESPACE [2]
+spec:
+  storageClassName: SOME-OBJECT-STORE-STORAGE-CLASS [3]
+  tenant: MY-TENANT [4]
+```
+
+1. `name` of the ObjectBucketClaim. Composes part of the bucket name and ConfigMap name
+1. `namespace` of the ObjectBucketClaim. Composes part of the bucket name. Determines the namespace of the ConfigMap
+1. `storageClassName` is used to target the desired Object Store.  Used by the operator to get the Object Store service URL.
+1. `tenant` allows users to define a tenant in an object store in order to namespace their buckets and access keys. 
+
+### OBC Custom Resource (Operator Updated)
+
+```yaml
+apiVersion: rook.io/v1
+kind: ObjectBucketClaim
+metadata:
+  name: MY-BUCKET-1
+  namespace: USER-NAMESPACE
+  ownerReferences: [1]
+  - name: CEPH-CLUSTER
+    ...
+spec:
+  storageClassName: SOME-OBJECT-STORE-STORAGE-CLASS
+status:
+  phase: {"pending", "bound", "error"}  [2]
+  objectBucketRef: objectReference{}  [3]
+  configMapRef: objectReference{}  [4]
+  secretRef: objectReference{}  [5]
+```
+
+1. `ownerReference` sets the ObjectBucketClaim as a child of the Ceph Cluster.  Cluster deletion causes the deletion of the OBC 
+1. `phase` 3 possible phases of bucket creation, mutually exclusive
+    - `pending`
+    - `bound`
+    - `error`
+1. `objectBucketRef` is an objectReference to the bound ObjectBucket 
+1. `configMapRef` is an objectReference to the generated ConfigMap 
+1. `secretRef` is an objectReference to the generated Secret
+
+### OB Custom Resource Definition
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: objectbucket.rook.io
+spec:
+  group: rook.io
+  version: v1
+  scope: namespaced
+  names:
+      kind: ObjectBucket
+      listKind: ObjectBucketList
+      singular: objectbucket
+      plural: objectbucket
+```
+
+### OB Custom Resource
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
```

Comment:
should this be `apiVersion: rook.io/v1`?

Reply:
This should have been changed when the doc was updated to be ceph-specific.   That said, the API isn't in the rook project, so it should be named accordingly.  Maybe `buckets.s3.io/v1alpha1`?  I'm open to suggestions.

Reply:
do we want `object` or `objectstore` or `objectbuckets` in the name or something to clearly spell out that this is object storage - guess `buckets` imply that?

Reply:
@screeley44 thanks, that's a good idea (I'm terrible at naming things)